File: logbook-spring-webflux/src/main/java/org/zalando/logbook/spring/webflux/ClientResponse.java
package org.zalando.logbook.spring.webflux;

import lombok.AllArgsConstructor;
import org.springframework.util.MimeType;
import org.zalando.logbook.HttpHeaders;
import org.zalando.logbook.HttpResponse;
import org.zalando.logbook.Origin;

import javax.annotation.Nullable;
import java.nio.charset.Charset;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicReference;

import static java.nio.charset.StandardCharsets.UTF_8;

@AllArgsConstructor
final class ClientResponse implements HttpResponse {

    private final org.springframework.web.reactive.function.client.ClientResponse response;

    private final AtomicReference<State> state = new AtomicReference<>(new State.Unbuffered());

    @Override
    public int getStatus() {
        return response.statusCode().value();
    }

    @Override
    public Origin getOrigin() {
        return Origin.REMOTE;
    }

    @Override
    public HttpHeaders getHeaders() {
        return HttpHeaders.of(response.headers().asHttpHeaders());
    }

    @Nullable
    @Override
    public String getContentType() {
        return Optional.ofNullable(response.headers().asHttpHeaders().getContentType())
                .map(MimeType::toString)
                .orElse(null);
    }

    @Override
    public Charset getCharset() {
        return Optional.ofNullable(response.headers().asHttpHeaders().getContentType())
                .map(MimeType::getCharset)
                .orElse(UTF_8);
    }

    @Override
    public HttpResponse withBody() {
        state.updateAndGet(State::with);
        return this;
    }

    @Override
    public HttpResponse withoutBody() {
        state.updateAndGet(State::without);
        return this;
    }

    boolean shouldBuffer() {
        return state.get() instanceof State.Offering;
    }

    void buffer(byte[] message) {
        state.updateAndGet(s -> s.buffer(message));
    }

    @Override
    public byte[] getBody() {
        return state.get().getBody();
    }
}

File: logbook-ktor-client/src/main/kotlin/org/zalando/logbook/client/ClientResponse.kt
@file:Suppress(
    "SimpleRedundantLet" // jacoco workaround
)

package org.zalando.logbook.client

import io.ktor.http.*
import io.ktor.util.*
import org.zalando.logbook.HttpHeaders
import org.zalando.logbook.HttpResponse
import org.zalando.logbook.Origin
import org.zalando.logbook.common.State
import java.nio.charset.Charset
import java.util.concurrent.atomic.AtomicReference
import kotlin.text.Charsets.UTF_8
import io.ktor.client.statement.HttpResponse as KtorResponse


internal class ClientResponse(
    private val response: KtorResponse
) : HttpResponse {
    private val state: AtomicReference<State> = AtomicReference(State.Unbuffered)

    override fun getProtocolVersion(): String = response.version.toString()
    override fun getOrigin(): Origin = Origin.REMOTE
    override fun getHeaders(): HttpHeaders = HttpHeaders.of(response.headers.toMap())
    override fun getContentType(): String? = response.contentType()?.let { it.toString().substringBefore(";") }
    override fun getCharset(): Charset = response.charset() ?: UTF_8
    override fun getStatus(): Int = response.status.value
    override fun withBody(): HttpResponse = apply { state.updateAndGet { it.with() } }
    override fun withoutBody(): HttpResponse = apply { state.updateAndGet { it.without() } }
    override fun getBody(): ByteArray = state.get().body
    internal fun buffer(bytes: ByteArray) = state.updateAndGet { it.buffer(bytes) }
    internal fun shouldBuffer(): Boolean = state.get() is State.Offering
}